File: HW8-JavaTetrisGui/HW8-JavaTetrisGui/src/tetris/Tetromino.java
package tetris;

/**
 * Class for Tetromino blocks
 * @author BeyzaAcar
 */


public class Tetromino {
    private char[][] block;
    private char block_letter;
    private int shift_right;
    private int shift_down;

    public Tetromino()
    {
        block = new char[4][4];
        for (int i = 0; i < 4; i++)
        {
            for (int j = 0; j < 4; j++)
            {
                block[i][j] = ' ';
            }
        }
    }

    /*
    public Tetromino(Tetromino obj)
    {
        for (int i = 0; i < 4; i++)
        {
            for (int j = 0; j < 4; j++)
            {
                block[i][j] = obj.getBlock()[i][j];
            }
        }
    }
    */

    /**
     * getter for block_letter
     * @return block_letter
     */
    public char getBlock_letter()
    {
        return block_letter;
    }

    /**
     * Changes the block according to the block_type
     * @param block_type type of the tetromino block
     */
    public void setData(char block_type) {
        shift_right = 0;
        shift_down = 0;

        for (int i = 0; i < 4; i++) {
            for (int j = 0; j < 4; j++) {
                block[i][j] = ' ';
            }
        }
        block_letter = block_type;

        if(block_type == 'I')
        {
            block[0][0] = 'x';
            block[0][1] = 'x';
            block[0][2] = 'x';
            block[0][3] = 'x';
        }
        else if(block_type == 'O')
        {
            block[2][0] = 'x';
            block[2][1] = 'x';
            block[3][0] = 'x';
            block[3][1] = 'x';
        }
        else if(block_type == 'T')
        {
            block[2][0] = 'x';
            block[2][1] = 'x';
            block[2][2] = 'x';
            block[3][1] = 'x';
        }
        else if(block_type == 'J')
        {
            block[1][1] = 'x';
            block[2][1] = 'x';
            block[3][0] = 'x';
            block[3][1] = 'x';
        }
        else if(block_type == 'L')
        {
            block[1][0] = 'x';
            block[2][0] = 'x';
            block[3][0] = 'x';
            block[3][1] = 'x';
        }
        else if(block_type == 'S')
        {
            block[2][1] = 'x';
            block[2][2] = 'x';
            block[3][0] = 'x';
            block[3][1] = 'x';
        }
        else if(block_type == 'Z')
        {
            block[2][0] = 'x';
            block[2][1] = 'x';
            block[3][1] = 'x';
            block[3][2] = 'x';
        }
        System.out.println("set datadansonra blok : ");
        for(int i = 0;i<4;i++)
        {
            for(int j = 0;j<4;j++)
            {
                System.out.print(block[i][j]);
            }
            System.out.println();
        }

    }
    /**
    *Sets the shift_right value
     */
    public void set_shift_right(int shift_right)
    {
        this.shift_right = shift_right;
    }

    /**
    *Allows us to access the shift_right variable.
     */
    public int  getShift_right()
    {
        return shift_right;
    }

    /**
     *Sets the shift_down value
     */
    public void setShift_down(int shift_down)
    {
        this.shift_down = shift_down;
    }

    /**
     *Allows us to access the shift_right variable.
     */
    public int getShift_down()
    {
        return shift_down;
    }
    
    /**
     * Moves the tetromino piece to the lower left corner of its own matrix.(4*4)
     */
    public boolean boslukSilme()
    {
        int flag1 = 0, flag2 = 0;
        int i;
        int j;
        for (i = 0; i < 4; i++)
        {
            if (block[i][0] == ' ') flag1++; /*leftside space*/
            if (block[3][i] == ' ') flag2++; /*downside space*/
        }
        if (flag2 == 4)
        {
            /*shifts down the block*/
            block[3] = block[2];
            block[2] = block[1];
            block[1] = block[0];
            block[0] = new char[]{' ', ' ', ' ', ' '};
            return true;
        }
        if (flag1 == 4)
        {
            for (i = 0; i < 4; i++)
            {
                for (j = 0; j < 3; j++)
                    block[i][j] = block[i][j + 1];
                block[i][3] = ' ';
            }
            return true;
        }
        return false;
    }

    /**
     * Rotates the tetromino block randomly
     */
    public void rotate(char dir)
    {
        char[][] temp_block = new char[4][4];
        int i,j,k;
        int dir_count = 1;
        if(dir == 'l') dir_count = 3;
        for(k=0;k<dir_count;k++)
        {
            for(i=0;i<4;i++)
            {
                for(j=3;j>=0;j--)
                {
                    temp_block[i][3-j] = block[j][i];
                }
            }
            for(i=0;i<4;i++)
            {
                for(j=0;j<4;j++)
                {
                    block[i][j] = temp_block[i][j];
                }
            }
        }
        while(boslukSilme() == true);
    }

    /**
     * getter for block
     * @return block as a char matrix
     */
    public char[][] getBlock() {
        return block;
    }
}

File: HW5-TemplateClass/HW5-TemplateClass/Tetromino.cpp
#include <iostream>
#include "Tetromino.hpp"

using namespace std;

Tetromino::Tetromino()
{
    int i,j;
    block = new char*[4];
    for(i=0;i<4;i++)
    {
        block[i] = new char[4];
        for(j=0;j<4;j++)
        {
            block[i][j] = ' ';
        }
    }
    dir_count = 0;
}
Tetromino::~Tetromino()
{
    int i;
    for(i=0;i<4;i++)
    {
        delete[] block[i];
    }
    delete[] block;
}
Tetromino::Tetromino(const Tetromino& obj)
{
    int i,j;
    for(i=0;i<4;i++)
    {
        for(j=0;j<4;j++)
        {
            block[i][j] = obj.get_block()[i][j];
        }
    }
}
bool Tetromino::bosluk_silme()
{
    int i, flag1 = 0, flag2 = 0;
    for(i = 0;i<4;i++)
    {
        if(block[i][0] == ' ')
            flag1++;
        if(block[3][i] == ' ')
            flag2++;
    }
    if(flag1 == 4)
    {
        for(i=0;i<3;i++)
        {
            block[0][i] = block[1][i];
            block[1][i] = block[2][i];
            block[2][i] = block[3][i];
            block[3][i] = ' ';
        }
        return true;
    } 
    if(flag2 == 4)
    {
        for(i=0;i<3;i++)
        {
            block[3][i] = block[2][i];
            block[2][i] = block[1][i];
            block[1][i] = block[0][i];
            block[0][i] = ' ';
        }
        return true;
    }
    return false;
}
void Tetromino::rotate()
{
    vector<vector<char>> temp(4);
    int i, j;
    for(j=0;j<=3;j++)
    {
        for(i=3;i>=0;i--)
        {
            temp[j].push_back(block[i][j]);
        }
    }
    for(i=0;i<4;i++)
    {
        for(j=0;j<4;j++)
        {
            block[i][j] = temp[i][j];
        }
    }
    while(bosluk_silme() == true);
}
char **Tetromino::get_block() const
{
    return block;
}
void Tetromino::set_data(const int dir_c, const char direction, const tetro block_type)
{
    dir_count = dir_c;
    dir = direction;
    int i,j;
    
    for(i=0;i<4;i++)
    {
        for(j=0;j<4;j++)
        {
            block[i][j] = ' ';
        }
    }

    if(block_type == tetro::I)
    {
        block[0][0] = 'I';
        block[0][1] = 'I';
        block[0][2] = 'I';
        block[0][3] = 'I';
    }
    else if(block_type == tetro::O)
    {
        block[2][0] = 'O';
        block[2][1] = 'O';
        block[3][0] = 'O';
        block[3][1] = 'O';
    }
    else if(block_type == tetro::T)
    {
        block[2][0] = 'T';
        block[2][1] = 'T';
        block[2][2] = 'T';
        block[3][1] = 'T';
    }
    else if(block_type == tetro::J)
    {
        block[1][1] = 'J';
        block[2][1] = 'J';
        block[3][0] = 'J';
        block[3][1] = 'J';
    }
    else if(block_type == tetro::L)
    {
        block[1][0] = 'L';
        block[2][0] = 'L';
        block[3][0] = 'L';
        block[3][1] = 'L';
    }
    else if(block_type == tetro::S)
    {
        block[2][1] = 'S';
        block[2][2] = 'S';
        block[3][0] = 'S';
        block[3][1] = 'S';
    }
    else if(block_type == tetro::Z)
    {
        block[2][0] = 'Z';
        block[2][1] = 'Z';
        block[3][1] = 'Z';
        block[3][2] = 'Z';
    }
    dir_count = dir_count%4;
    if(dir == 'l') dir_count = 4-dir_count;
    for(i=0;i<dir_count;i++)
    {
        rotate();
    }
}
void Tetromino::set_block( char **temp_block)
{
    int i,j;
    for(i=0;i<4;i++)
    {
        for(j=0;j<4;j++)
        {
            block[i][j] = temp_block[i][j];
        }
    }
}
int Tetromino::get_dir_count() const
{
    return dir_count;
}
char Tetromino::get_dir() const
{
    return dir;
}
void Tetromino::set_dir_count(const int a)
{
    dir_count = a;
}
void Tetromino::set_dir(const char a)
{
    dir = a;
}

